### DISCUSSION TITLE
Excel 2016 ExcelAsyncHandle crash

### DISCUSSION THREAD

**Discussion by aytactaner** [Excel-DNA/ExcelDna#564] [Q&A]
Hi,

The document which is created in Office365 is crashing Excel application while opening in Excel 2016 (Microsoft® Excel® 2016 MSO (Version 2301 Build 16.0.16026.20002) 32-bit ) version. Please find attached the document and add-in project. Do you have any idea ?
Thank you.

[document.xlsx](https://github.com/Excel-DNA/ExcelDna/files/10579650/document.xlsx)
[Project.zip](https://github.com/Excel-DNA/ExcelDna/files/10579648/Project.zip)

**Comment by govert** [comment]
Hi @aytactaner - can you please confirm whether this is a problem with the latest Excel-DNA v1.6.1-beta3?

**Comment by aytactaner** [comment]
Hi Govert,
I've tried using Excel-DNA v1.6.1-beta3. Error still exists. Excel crashed with "Access Violation Error".

[Project-1.6.1-beta3.zip](https://github.com/Excel-DNA/ExcelDna/files/10580621/Project-1.6.1-beta3.zip)

**Comment by govert** [comment]
@aytactaner I've had a look at your project, but not had a chance to check under Excel 2016. But I see you are returning a 1x1 array from the native async function. I wonder if the problem you have might relate to the fact that Excel 365 has 'dynamic arrays' support and the Excel 2016 does not. Could you perhaps test with a function that only returns a value directly, not in an array?

**Comment by govert** [comment]
And does it only crash when you create or save the file with new Excel and then open with old Excel?
If so, can you make new file with old Excel that does not crash, and we can compare the two Excel files.

Or, does the function always crash with old Excel, even when you create a new file.

**Comment by aytactaner** [comment]
Even I create new file in Excel 2016 or 2019 it crashes. But I realized something new. Only crashes if one of cells refereces to another cell which returns array (both of are native async). If I remove that dependency I works without any problem.

**Comment by aytactaner** [comment]
Also, if I ExcelAsyncHandle.SetResult immediately instead of Task.Run it works.

**Comment by govert** [comment]
OK - I got VM with Excel 2016 running.
I can't reproduce a problem though.
When I call either of your functions `RasLastDay` or `RasRatio` I just get #VALUE.

It feels like native async can't handle arrays before Excel 365.
But you see something else ?

* Can you write some instructions for what I can type into Excel to see your problem?
* Can you confirm whether you are targeting .NET Framework or .NET 6?

**Reply by aytactaner** [reply]
Hi Govert,
I captured a screenshot video. I hope you can reproduce the issue. I can provide additional information if you need.
Thank you.

https://youtu.be/yDUhDq3lhmU
[document3.xlsx](https://github.com/Excel-DNA/ExcelDna/files/10720920/document3.xlsx)

**Comment by govert** [comment]
* Do you have any other add-ins loaded in Excel?
* Does the crash also happen when you type in the formula into a new workbook?
* Does the crash also happen when you use no array formulas, but just simple (async) functions that return a single value?
* I see that you are running 32-bit Excel - I only have access to the 32-bit version at the moment, and don't know if that makes a difference.

**Reply by aytactaner** [reply]
- I don't have any other addins loaded. 
- It crashes only when I opened a workbook saved in Excel 365. I can write new formulas without any issue.
- It doesn't crash when using non-array formulas
- I've tried in Excel 2019 - 64 bit. Result is same, it crashes.

I think that Excel 365 mark all external addin functions as array while saving into xlsx file. While openning that file in 2016 or 2019 and if functions are native async Excel could not handle calculation chain in correct order for that cells. Is it possible ? What do you think ?

**Comment by govert** [comment]
You say
>It crashes only when I opened a workbook saved in Excel 365. I can write new formulas without any issue.

But earlier you said
> Even I create new file in Excel 2016 or 2019 it crashes. 

Can you clarify?


I suspect this is an Excel bug, where the combination of native async UDF and array results are crashing Excel 2016.
Do you have to support array-returning UDFs and also Excel 2016?

If this is an Excel bug, you could check the version of Excel in the function and only return an array if the version supports Dynamic Arrays. 

```c#
  public static class UtilityFunctions
    {
        static bool? _supportsDynamicArrays;
        [ExcelFunction(IsHidden = true)]
        public static bool dnaSupportsDynamicArrays()
        {
            if (!_supportsDynamicArrays.HasValue)
            {
                try
                {
                    var result = XlCall.Excel(614, new object[] { 1 }, new object[] { true });
                    _supportsDynamicArrays = true;
                }
                catch
                {
                    _supportsDynamicArrays = false;
                }
            }
            return _supportsDynamicArrays.Value;
        }
    }
```

**Comment by aytactaner** [comment]
I mean I can write new scalar formulas, not arrays in 2016. If I make them as array using Control+Shift+Enter and it referenced from another cell Excel crashes.

I have about 2.5k functions and a few of them are have to return array. 

I don't think dynamic array support checking can be a workaround because it crashes even if I do not return nothing. Following function also crashes if referenced from another cell. Because Excel expects array from that function.

 ```
public static void RasLastDay(object value1, ExcelAsyncHandle asyncHandle)
{
    //Do nothing
}
```

**Comment by govert** [comment]
OK - I can confirm that I can get 64-bit Excel 2016 version 2104 to crash too, using the combination of native async array function calls.

```
Faulting application name: EXCEL.EXE, version: 16.0.13929.20296, time stamp: 0x6088c17c
Faulting module name: EXCEL.EXE, version: 16.0.13929.20296, time stamp: 0x6088c17c
Exception code: 0xc0000005
Fault offset: 0x00000000023aab6f
Faulting process id: 0x1d4c
Faulting application start time: 0x01d93fa170ac0767
Faulting application path: C:\Program Files\Microsoft Office\Root\Office16\EXCEL.EXE
Faulting module path: C:\Program Files\Microsoft Office\Root\Office16\EXCEL.EXE
Report Id: 7c3ef3f8-85a7-409c-8b5b-96d6489a243b
Faulting package full name: 
Faulting package-relative application ID: 
```

It might be worth pressing the Feedback -> Send a Frown button to report the problem.

The next step would be to reproduce this with a native code .xll add-in. If this has the same bahaviour, we know that it is an Excel bug and not related to Excel-DNA. I'm not sure when I'll get a chance.

Then, if this is an Excel bug, it would seem very unlikely that they would fix it, unless you are a very important account for them. Do you have any direct support channel to Microsoft through an account manager?

**Reply by aytactaner** [reply]
I also agree with you, this is an Excel bug. Thank you for your support. I will try my chance. 

Regards.
